### ISSUE TITLE
Start using React's new context api

### ISSUE THREAD

**Issue by vejersele** [vejersele/react-stateful-component#114]

**Comment by girvo** [comment]
Would be absolutely wonderful. I'm also stoked to see I'm not the only crazy person who wants `reducerComponent` in Flow haha

**Comment by vejersele** [comment]
Thanks for your comment! I've been looking into this, but it seems that the new context api doesn't provide a way to access context from a react component's lifecycle hooks or constructor. As a result I would need to introduce a wrapping component to grab the context and passed it on as a prop to the stateful component. Since that feels kind of bloated, I'm thinking about dropping the sideEffectProvider. Not sure yet though.

**Comment by girvo** [comment]
Yeah I believe they are deprecating some of the lifecycle methods, moving forward.

I think a single wrapping component wouldn't be too bad, but yeah personally I'd love for there to be no wrappers at all. I'm not 100% certain if this is doable, but I have a sketched-out `ReducerComponent` myself that uses render props and "steals" some of your code, to attempt to get rid of the wrapper. Not 100% happy with it yet, however.

Also, quick note: https://github.com/girvo/reducer-component-test

For the `render` method in `src/App.js` I had to add `key`'s to the children of `<div>` -- not sure why though! Normally React would handle that, but I'm guessing because we're using a render prop with a more custom system it's automatic key generation for children fails?

**Comment by vejersele** [comment]
I've just merged a PR which contains the necessary changes for this library to make use of the new context api. Seems to be working fine, with the added advantage that <SideEffectProvider> is no longer necessary if you need the default behaviour (executing side effects). 

While unit testing you can still use the mock provider to intercept side effects if you need to. These changes aren't released yet, I need to update the docs first and do some testing with older versions of react. Public API should stay the same though.

To come back to your issue with react complaining about key's ... Could this be related to your build setup? I'm not familiar with rollup, but I found some issues in the react issue tracker that seemed similar to what you're experiencing. They're not about rollup though (https://github.com/facebook/react/issues/11537).

I've never experienced this during testing, and I would be surprised if it would be caused by this library. You might want to try it out with Create React App (https://github.com/facebook/create-react-app)
